### PR TITLE
scanner: Create and store file listings for each resolved provenance

### PR DESCRIPTION
**Pull request by fviernau**
Create and store the file listings `(path, sha1)` for each resolved provenance. 
These can be consumed in future iterations (see #6945) for improving the generated SBOMs.

Resolves: #6943.

I've scanned the HEAD of the main branch of npm mime-types which resolved to 335 provenances. The total size of all file listings in the Postgres database is 5605372 bytes. So, about 16.4kb per file listing in average.

BREAKING CHANGE: One does not have to, but should configure a storage backend in `~/.ort/config/config.yml`.
                                      If one doesn't do so, ORT falls back to storing the file listings under `~/.ort` which may not be
                                      desired. See the modified `reference.yml` for how the new configuration feature looks like.

### CODE REVIEW

File: scanner/src/main/kotlin/utils/FileListingResolver.kt [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+                )
+                child.isDirectory -> if (!IGNORED_DIRECTORY_MATCHER.matches(path)) {
+                    queue += child
+                } else {
+                    println("ignoring: $path")
+                }
+                else -> return@forEach
+            }
+        }
+    }
```
Comment (sschuberth) [inline scanner/src/main/kotlin/utils/FileListingResolver.kt:102]: Why not simply use `File.walk()`instead of a custom queue implementation?

File: model/src/main/kotlin/config/FileListingStorageConfiguration.kt [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+     * Configuration of the [PostgresProvenanceFileStorage] used for storing the file listings.
+     */
+    val postgresStorage: PostgresStorageConfiguration? = null
+) {
+    private companion object : Logging
+
+    fun createStorage(): ProvenanceFileStorage =
+        when {
+            fileStorage != null -> {
+                if (postgresStorage != null) {
```
Comment (sschuberth) [inline model/src/main/kotlin/config/FileListingStorageConfiguration.kt:51]: Personally, I'd prefer a more "direct" check for this in an `init`-block like done in `FileArchiverConfiguration`.

Comment (sschuberth) [reply model/src/main/kotlin/config/FileListingStorageConfiguration.kt:51]: In general, I wonder whether this should resemble `FileArchiverConfiguration` more, also regarding the extension function for the creation, and the general `enable` flag. I'm curious about @mnonnenmacher's opinion here.

File: scanner/src/main/kotlin/Scanner.kt [hunk 3]
```diff
@@ -577,6 +587,41 @@ class Scanner(
...
+
+        val duration = measureTime {
+            val previousIndex = AtomicInteger(0)
+
+            provenancesMissingFileListings.parallelStream().forEach { provenance ->
+                val index = previousIndex.incrementAndGet()
+
+                logger.info { "Creating file listing for provenance $index of ${provenancesMissingFileListings.size}." }
+
+                kotlin.runCatching {
```
Comment (sschuberth) [inline scanner/src/main/kotlin/Scanner.kt:604]: This should just be `runCatching`.

File: scanner/src/main/kotlin/Scanner.kt [hunk 4]
```diff
@@ -577,6 +587,41 @@ class Scanner(
...
+                    fileListingResolver.resolve(provenance)
+                }.onFailure {
+                    idsByProvenance.getValue(provenance).forEach { id ->
+                        controller.addIssue(
+                            id,
+                            Issue(
+                                source = "Downloader",
+                                message = "Could not create file listing for " +
+                                        "'${id.toCoordinates()}': ${it.collectMessages()}",
+                                severity = Severity.ERROR
```
Comment (sschuberth) [inline scanner/src/main/kotlin/Scanner.kt:614]: This is the default and can be omitted.

File: scanner/src/main/kotlin/Scanner.kt [hunk 5]
```diff
@@ -577,6 +587,41 @@ class Scanner(
...
+                            )
+                        )
+                    }
+                }
+            }
+        }
+
+        logger.info("Created file listing for ${provenancesMissingFileListings.size} provenances in $duration.")
```
Comment (sschuberth) [inline scanner/src/main/kotlin/Scanner.kt:622]: Use `{}` for lazy evaluation.

File: scanner/src/main/kotlin/Scanner.kt [hunk 6]
```diff
@@ -577,6 +587,41 @@ class Scanner(
...
+    private fun createMissingFileListings(controller: ScanController) {
+        val idsByProvenance = controller.getIdsByProvenance()
+        val provenancesMissingFileListings = idsByProvenance.keys.filterNot { fileListingResolver.has(it) }
+
+        logger.info { "Creating file listings for ${provenancesMissingFileListings.size} provenances." }
+
+        val duration = measureTime {
+            val previousIndex = AtomicInteger(0)
+
+            provenancesMissingFileListings.parallelStream().forEach { provenance ->
```
Comment (sschuberth) [inline scanner/src/main/kotlin/Scanner.kt:599]: As `parallelStream()` is limited to JVM targets, I'd prefer a more Kotlin-style approach with coroutines, similar to like recently done for the downloader CLI:

https://github.com/oss-review-toolkit/ort/blob/eb88bd776c0accb75f3741aaf398d858c4428739/cli/src/main/kotlin/commands/DownloaderCommand.kt#L312-L321

That'd also allow you to get rid of the manual counting with `AtomicInteger`.

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: scanner/src/main/kotlin/FileListing.kt [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        val sha1: String
+    )
+
+    init {
+        val duplicates = files.getDuplicates { it.path }.keys
+
+        require(duplicates.isEmpty()) {
+            "The file listing contains duplicate paths which is not allowed: ${duplicates.joinToString()}."
+        }
+    }
```
[the remark this answers is not in the document]
Comment (fviernau) [reply scanner/src/main/kotlin/FileListing.kt:55]: > Have you considered using a map with `path` as the key and hash(es) (or other file properties) as values to avoid the need to this check?

Yes I considered that. I didn't / don't find this check problematic - I thought the data structure as-is would be a tad more speaking.

File: scanner/src/main/kotlin/utils/FileListingResolver.kt (hunk 1 above)
Comment (fviernau) [reply scanner/src/main/kotlin/utils/FileListingResolver.kt:102]: I changed it to using `walk()`.

File: model/src/main/kotlin/config/FileListingStorageConfiguration.kt (hunk 2 above)
Comment (fviernau) [reply model/src/main/kotlin/config/FileListingStorageConfiguration.kt:51]: > Personally, I'd prefer a more "direct" check for this in an `init`-block like done in `FileArchiverConfiguration`.

That mentioned init block is executed when ort config is loaded. I thought it would be better to only log it exactly when the storage gets created. I still think it's no a bad idea, but I do not have such a strong opinion on that, so I could move it to `init` if you want.

Comment (fviernau) [reply model/src/main/kotlin/config/FileListingStorageConfiguration.kt:51]: > In general, I wonder whether this should resemble `FileArchiverConfiguration` more, also regarding the extension function for the creation, and the general `enable` flag. I'm curious about @mnonnenmacher's opinion here.

It would have been nice to hear some rationale for this. Anyhow, I deliberately have made it mandatory so that we can count on the file listings to be present for simplicity. Making the file listing optional I prefer to not be part of the (initial) implementation.

Comment (mnonnenmacher) [reply model/src/main/kotlin/config/FileListingStorageConfiguration.kt:51]: > In general, I wonder whether this should resemble FileArchiverConfiguration more, also regarding the extension function for the creation, and the general enable flag. I'm curious about @mnonnenmacher's opinion here.

I need to have a look at the code tomorrow before I can answer that.

> Anyhow, I deliberately have made it mandatory so that we can count on the file listings to be present for simplicity. Making the file listing optional I prefer to not be part of the (initial) implementation.

IMO the file lists should be mandatory, this makes it easier to build the features that depend on it.

> One does not have to, but should configure a storage backend in ~/.ort/config/config.yml. If one doesn't do so, ORT falls back to storing the file listings under ~/.ort which may not be desired.

Off topic, but this reminds me of #5516 which would prevent that we have yet another place to configure the same database connection. I wonder if we really need the flexibility to configure all storage types for the scanner separately, or if one central configuration to configure either a file storage or a Postgres storage would be good enough.

Comment (fviernau) [reply model/src/main/kotlin/config/FileListingStorageConfiguration.kt:51]: @mnonnenmacher I could add #5516 to my epic, e.g. as last clean-up ticket. Ok?

File: scanner/src/main/kotlin/FileListing.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (sschuberth) [reply scanner/src/main/kotlin/FileListing.kt:55]: > Do you think a map would be better?

I'm unsure myself, esp. when it comes to how the serialization looks, but that's probably anyway unimportant here. I was more considering the case when more properties, like a Git-style SHA1, would be added. Maybe if would be nicer to have paths to files and files properties more clearly "separated" via key and values.

[the remark this answers is not in the document]
Comment (sschuberth) [reply scanner/src/main/kotlin/FileListing.kt:55]: Speaking of more properties to add, have you also considered to use (a set of) `Hash` classes instead of `String`s (and ensure that at least an entry with `HashAlgorithm.SHA1` is present)?

File: model/src/main/kotlin/config/FileListingStorageConfiguration.kt (hunk 2 above)
Comment (sschuberth) [reply model/src/main/kotlin/config/FileListingStorageConfiguration.kt:51]: > > In general, I wonder whether this should resemble `FileArchiverConfiguration` more, also regarding the extension function for the creation, and the general `enable` flag. I'm curious about @mnonnenmacher's opinion here.
> 
> It would have been nice to hear some rationale for this.

IMO it's wrong that `FileListingStorageConfiguration` knows how to create `ProvenanceFileStorage`. It should be the opposite: `ProvenanceFileStorage` knowns how create itself given a `FileListingStorageConfiguration`. However, the latter is hard to implement cleanly as we can have either a `FileProvenanceFileStorage` or a `PostgresProvenanceFileStorage`. So we need to move creation to a "higher level" / common place, which leads us back to `FileListingStorageConfiguration`. But *if* we do it in `FileListingStorageConfiguration`, I find it to be more clean if it's separated from the data class via an extension function like done in `FileArchiverConfiguration`.

Comment (sschuberth) [reply model/src/main/kotlin/config/FileListingStorageConfiguration.kt:51]: > > Anyhow, I deliberately have made it mandatory so that we can count on the file listings to be present for simplicity. Making the file listing optional I prefer to not be part of the (initial) implementation.
> 
> IMO the file lists should be mandatory, this makes it easier to build the features that depend on it.

Why is it okay to have an optional file storage for license files, but it's not okay to make the storage for listings optional?

If the argument is simplicity, we should also make the license file storage mandatory, as anyway pretty much everybody seems to agree that the original wording of the license files needs to be retained.

File: scanner/src/main/kotlin/FileListing.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (fviernau) [reply scanner/src/main/kotlin/FileListing.kt:55]: > I'm unsure myself, esp. when it comes to how the serialization looks, but that's probably anyway unimportant here. I was more considering the case when more properties, like a Git-style SHA1, would be added. Maybe if would be nicer to have paths to files and files properties more clearly "separated" via key and values.

I believe a dedicated property is a "clear" separation already.
Adding further properties is out of scope of the this PR and the whole epic.
(This can be refactored quite easily if needed in the future)

[the remark this answers is not in the document]
Comment (fviernau) [reply scanner/src/main/kotlin/FileListing.kt:55]: > Speaking of more properties to add, have you also considered to use (a set of) `Hash` classes instead of `String`s (and ensure that at least an entry with `HashAlgorithm.SHA1` is present)?

No, this is not needed for an MVP.

File: model/src/main/kotlin/config/FileListingStorageConfiguration.kt (hunk 2 above)
Comment (fviernau) [reply model/src/main/kotlin/config/FileListingStorageConfiguration.kt:51]: > But if we do it in FileListingStorageConfiguration, I find it to be more clean if it's separated from the data class

I'm not following the arguments "separation" and "more clean, to me it's not obvious  

1. why an extension function is more clean
2. in what way the degree of separation is increased.

I'll change it anyway

File: scanner/src/main/kotlin/FileListing.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (sschuberth) [reply scanner/src/main/kotlin/FileListing.kt:55]: > I believe a dedicated property is a "clear" separation already.

That depends on whether you regard the path to be a property of the file, like the SHA1. Or if you believe the path to be an identifier for the file, and the file (contents) have properties separate from that, like hashes, modification time, etc.

> Adding further properties is out of scope of the this PR and the whole epic.
> (This can be refactored quite easily if needed in the future)

It's certainly needed, and it also corresponds to what we discussed some time ago: Any implementation of this epic must be prepared for adding additional hashes (like the mentioned Git-style SHA1, for Software Heritage interoperability).

> No, this is not needed for an MVP IMO.

Obviously it's not technically needed. The question is whether it makes sense / is prescient to already implement it like that. Or phrased differently, why does it make sense *not* to use the `Hash` class here, since we already use it elsewhere for serialization.

File: model/src/main/kotlin/config/FileListingStorageConfiguration.kt (hunk 2 above)
Comment (sschuberth) [reply model/src/main/kotlin/config/FileListingStorageConfiguration.kt]: > > we should also make the license file storage mandatory, as anyway pretty much everybody seems to agree that the original wording of the license files needs to be retained.
> 
> Maybe, but can we leave this discussion out of scope here?

Sorry, no. I'm not saying we must already implement this now, but we should be allowed to discuss it, as it obviously has an impact on this PR.

I have a problem with bending arguments and changing minds every time it suites somebody's needs, just because there is pressure to implement a feature. Out of a sudden "simplicity" is an argument here, whereas in the past we were forced to add complexity by hiding several (IMO uncritical) features behind feature flags. I just miss the consistency and equality in the line of though here.

Comment (sschuberth) [reply model/src/main/kotlin/config/FileListingStorageConfiguration.kt]: > I'm not following the arguments "separation" and "more clean, to me it's not obvious
> 
>     1. why an extension function is more clean
>     2. in what way the degree of separation is increased.

My answer to both is: Because the implementation is not part of the class body anymore. (I.e. it's more of a "visual" separation that expresses the intent better, as that's probably the best we can do here.)

File: scanner/src/main/kotlin/FileListing.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (fviernau) [reply scanner/src/main/kotlin/FileListing.kt:55]: > It's certainly needed, and it also corresponds to what we discussed some time ago: Any implementation of this epic must be prepared for adding additional hashes (like the mentioned Git-style SHA1, for Software Heritage interoperability).

As said, it's easy as-is already to add further properties. So, yes it is prepared well enough IMO.

> Obviously it's not technically needed. The question is whether it makes sense / is prescient to already implement it like that. Or phrased differently, why does it make sense not to use the Hash class here, since we already use it elsewhere for serialization.

It does make sense to not do it because it is not needed in scope of my epic.

[the remark this answers is not in the document]
Comment (sschuberth) [reply scanner/src/main/kotlin/FileListing.kt:55]: > It does make sense to not do it because it is not needed in scope of my epic.

So is our policy now that everybody just strictly scratches their own itch, disregarding what other related use-cases others may have in the future, and leaving any additional work to them, even if it would not hurt to do it now? That's not the way at least I want to work on this project.

**Comment by fviernau** [comment]
Looking at above discussion I have a hunch it may be more efficient to come to a conclusion in a short call. @sschuberth @mnonnenmacher  would you be in for that?

File: scanner/src/main/kotlin/FileListing.kt [hunk 8]
```diff
@@ -0,0 +1,60 @@
+/*
```
Comment (fviernau) [inline scanner/src/main/kotlin/FileListing.kt:1]: commit-msg: Don't mention "scanned" but say "each resolved provenance". Same for PR description.

File: scanner/src/main/kotlin/FileListing.kt [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+import java.util.SortedSet
+
+import org.ossreviewtoolkit.scanner.FileListing.FileEntry
+import org.ossreviewtoolkit.utils.common.StringSortedSetConverter
+import org.ossreviewtoolkit.utils.common.getDuplicates
+
+/**
+ * The model to store file listings for provenances in a storage.
+ */
```
Comment (fviernau) [inline scanner/src/main/kotlin/FileListing.kt:33]: The model to store a file listing for a resolved provenance.

File: scanner/src/test/assets/expected-file-listing.yml [hunk 10]
```diff
@@ -0,0 +1,14 @@
+---
+ignored_directories:
```
Comment (fviernau) [inline scanner/src/test/assets/expected-file-listing.yml:2]: rename to : ignore_patterns

File: scanner/src/main/kotlin/FileListing.kt [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+internal data class FileListing(
+    /**
+     * A set of glob expressions that match directory paths which have been excluded from the file listing.
+     */
+    @JsonSerialize(converter = StringSortedSetConverter::class)
+    val ignoredDirectories: Set<String>,
+
+    @JsonSerialize(converter = FileEntrySortedSetConverter::class)
```
Comment (fviernau) [inline scanner/src/main/kotlin/FileListing.kt:41]: Add KDoc for `files`.

File: scanner/src/main/kotlin/provenance/ProvenanceDownloader.kt [hunk 12]
```diff
@@ -39,7 +39,7 @@ import org.ossreviewtoolkit.utils.ort.createOrtTempDir
 /**
  * An interface that provides functionality to download source code.
  */
-interface ProvenanceDownloader {
+fun interface ProvenanceDownloader {
```
Comment (fviernau) [inline scanner/src/main/kotlin/provenance/ProvenanceDownloader.kt:42]: Try if FakeProvenanceDownloader can use it. If yes, make separate commit.

File: model/src/main/kotlin/config/FileListingStorageConfiguration.kt [hunk 13]
```diff
@@ -0,0 +1,75 @@
+/*
```
Comment (fviernau) [inline model/src/main/kotlin/config/FileListingStorageConfiguration.kt:1]: commit-msg: scanned provenance -> resolved provenance

File: model/src/main/kotlin/config/FileListingStorageConfiguration.kt [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+        )
+        this?.postgresStorage != null -> PostgresProvenanceFileStorage(
+            dataSource = DatabaseUtils.createHikariDataSource(
+                config = postgresStorage.connection,
+                applicationNameSuffix = "file-listings"
+            ),
+            tableName = TABLE_NAME
+        )
+        else -> FileProvenanceFileStorage(
+            storage = LocalFileStorage(ortDataDirectory.resolve("scanner/provenance-file-listings")),
```
Comment (fviernau) [inline model/src/main/kotlin/config/FileListingStorageConfiguration.kt:72]: just: "scanner/file-listings"

File: scanner/src/main/kotlin/Scanner.kt [hunk 15]
```diff
@@ -577,6 +586,43 @@ class Scanner(
...
+                            fileListingResolver.resolve(provenance)
+                        }.onFailure {
+                            idsByProvenance.getValue(provenance).forEach { id ->
+                                controller.addIssue(
+                                    id,
+                                    Issue(
+                                        source = "Downloader",
+                                        message = "Could not create file listing for " +
+                                                "'${id.toCoordinates()}': ${it.collectMessages()}",
+                                        severity = Severity.ERROR
```
Comment (fviernau) [inline scanner/src/main/kotlin/Scanner.kt:613]: drop ` severity = Severity.ERROR`.

File: scanner/src/main/kotlin/Scanner.kt [hunk 16]
```diff
@@ -192,7 +201,7 @@ class Scanner(
                 }.onFailure {
                     controller.addProvenanceResolutionIssue(
                         pkg.id,
-                        Issue(source = TOOL_NAME, severity = Severity.ERROR, message = it.collectMessages())
+                        Issue(source = TOOL_NAME, message = it.collectMessages())
```
Comment (fviernau) [inline scanner/src/main/kotlin/Scanner.kt:204]: this is done in wrong commit.

File: scanner/src/main/kotlin/provenance/ProvenanceDownloader.kt (hunk 12 above)
Comment (fviernau) [reply scanner/src/main/kotlin/provenance/ProvenanceDownloader.kt:42]: The other implementation (FakeProvenanceDownloader) could be turned into a function which return an anonymous class / instance. However, this does not work with the spies used with that instance.

**Comment by fviernau** [comment]
@mnonnenmacher @sschuberth I've incorporated the changes we've agreed upon in our call yesterday.

File: scanner/src/test/kotlin/utils/FileListingResolverTest.kt [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+                    ".git/index",
+                    "LICENSE",
+                    "src/cli/main.cpp"
+                )
+            },
+        )
+
+        val fileListing = resolver.resolve(ArtifactProvenance(sourceArtifact = RemoteArtifact.EMPTY))
+
+        fileListing.toYaml() shouldBe File("src/test/assets/expected-file-listing.yml").readText()
```
Comment (mnonnenmacher) [inline scanner/src/test/kotlin/utils/FileListingResolverTest.kt:53]: It would be good to have some more fine grained test, e.g. to verify that the exclude patterns work as expected. I'm fine with adding them later.
If more tests are added, it might be better for the overview to inline the expected results in the tests instead of using separate files.

File: model/src/main/kotlin/config/FileListingStorageConfiguration.kt [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+    val postgresStorage: PostgresStorageConfiguration? = null
+) {
+    private companion object : Logging
+
+    init {
+        if (fileStorage != null && postgresStorage != null) {
+            FileListingStorageConfiguration.logger.warn {
+                "'fileStorage' and 'postgresStorage' are both configured but only one storage can be used. " +
+                        "Using 'fileStorage'."
+            }
```
Comment (mnonnenmacher) [inline model/src/main/kotlin/config/FileListingStorageConfiguration.kt:53]: Such warnings are easily overlooked, why not throw an exception?

File: scanner/src/main/kotlin/Scanner.kt (hunk 16 above)
Comment (mnonnenmacher) [reply scanner/src/main/kotlin/Scanner.kt:203]: What is the rationale for this? The default might change, but this would still be an error.

File: scanner/src/main/kotlin/ScanController.kt [hunk 19]
```diff
@@ -142,6 +142,18 @@ class ScanController(
...
+     * Return all provenances including sub-repositories associated with the identifiers of the packages they belong to.
+     */
+    fun getIdsByProvenance(): Map<KnownProvenance, Set<Identifier>> =
+        buildMap<_, MutableSet<Identifier>> {
+            getNestedProvenancesByPackage().forEach { (pkg, nestedProvenance) ->
+                nestedProvenance.getProvenances().forEach { provenance ->
+                    getOrPut(provenance) { mutableSetOf() } += pkg.id
+                }
+            }
+        }
```
Comment (mnonnenmacher) [inline scanner/src/main/kotlin/ScanController.kt:155]: Would `getPackagesConsolidatedByProvenance().mapValues { it.value.id }` create the same results with less code?

**Review by mnonnenmacher — CHANGES_REQUESTED** [review]

**Comment by mnonnenmacher** [comment]
> > What is the rationale for this? The default might change, but this would still be an error.
> 
> @mnonnenmacher the removal of the explicit `Severity.ERROR` happened on my request to align with the code base. If we cannot rely on the value of default arguments, it makes little sense to use default values at all...

Then I would argue for removing the default value, the issue severity is something I would like to see on the call side.

File: scanner/src/main/kotlin/Scanner.kt (hunk 16 above)
Comment (fviernau) [reply scanner/src/main/kotlin/Scanner.kt:203]: @sschuberth was asking in this review to not pass redundant default values. IIRC rationale was consistency.

File: scanner/src/test/kotlin/utils/FileListingResolverTest.kt (hunk 17 above)
Comment (fviernau) [reply scanner/src/test/kotlin/utils/FileListingResolverTest.kt:53]: I'll do it in a follow up PR then.

File: model/src/main/kotlin/config/FileListingStorageConfiguration.kt (hunk 18 above)
Comment (fviernau) [reply model/src/main/kotlin/config/FileListingStorageConfiguration.kt:53]: This is consistent with the FileArchiverConfiguration. There have been long discussions about this class in this PR, can we leave as-is now. This as-is IIUC what was agreed with @sschuberth .

**Comment by fviernau** [comment]
> Then I would argue for removing the default value, the issue severity is something I would like to see on the call side.

Are you fine with de-scoping this from this PR then. I don't like changing this back and forth. In my view this is now a different topic. Agreed?
